src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,393 @@
+     * @throws UnsupportedOperationException
+     *     if the derived keying material is not extractable
+     */
+    @Override
+    protected byte[] engineDeriveData(AlgorithmParameterSpec derivationSpec)
+        throws InvalidAlgorithmParameterException {
+        List<SecretKey> ikms;
+        List<SecretKey> salts;
+        byte[] inputKeyMaterial;
+        byte[] salt;
+        byte[] pseudoRandomKey;
+        byte[] info;
+        int length;
+        // A switch would be nicer, but we may need to backport this before
+        // JDK 17
+        // Also, JEP 305 came out in JDK 14, so we can't declare a variable
+        // in instanceof either
+        if (derivationSpec instanceof HKDFParameterSpec.Extract) {
+            HKDFParameterSpec.Extract anExtract =
+                (HKDFParameterSpec.Extract) derivationSpec;
+            ikms = anExtract.ikms();
+            salts = anExtract.salts();
+            // we should be able to combine both of the above Lists of key
+            // segments into one SecretKey object each, unless we were passed
+            // something bogus or an unexportable P11 key
+            try {
+                inputKeyMaterial = consolidateKeyMaterial(ikms);
+                salt = consolidateKeyMaterial(salts);
+            } catch (InvalidKeyException ike) {
+                throw (InvalidAlgorithmParameterException) new InvalidAlgorithmParameterException(
+                    "Issue encountered when combining ikm or salt values into"
+                    + " single keys").initCause(ike);
+            }
+            // perform extract
+            try {
+                return hkdfExtract(inputKeyMaterial, salt);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidAlgorithmParameterException(
+                    "an HKDF Extract could not be initialized with the given "
+                    + "key or salt material", ike);
+            } catch (NoSuchAlgorithmException nsae) {
+                // This is bubbling up from the getInstance of the Mac/Hmac.
+                // Since we're defining these values internally, it is unlikely.
+                throw new ProviderException(
+                    "could not instantiate a Mac with the provided algorithm",
+                    nsae);
+            }
+        } else if (derivationSpec instanceof HKDFParameterSpec.Expand) {
+            HKDFParameterSpec.Expand anExpand =
+                (HKDFParameterSpec.Expand) derivationSpec;
+            // set this value in the "if"
+            if ((pseudoRandomKey = anExpand.prk().getEncoded()) == null) {
+                throw new AssertionError(
+                    "PRK is required for HKDFParameterSpec.Expand");
+            }
+            // set this value in the "if"
+            if ((info = anExpand.info()) == null) {
+                info = new byte[0];
+            }
+            length = anExpand.length();
+            if (length > (hmacLen * 255)) {
```

Comment:
Since hmacLen is an int, it makes sense to make the multiplication a long multiplication to avoid potential risk of overflow?

```suggestion
            if (length > (hmacLen * 255L)) {
```